File: algorithm/sort/bubble_sort_2.c
#include <stdbool.h>

/**
 * 배열 내 자리 이동이 더 이상 발생하지 않으면
 * 정렬이 완료됐다는 의미라 for loop 종료
 * https://www.algodale.com/algorithms/bubble-sort/
 */
void bubble_sort_2(int *arr, const int size) {
    for (int i = size - 1; i > 0; i--) {
        bool is_switched = false;
        for (int j = 0; j < i; j++) {
            if (arr[j] > arr[j + 1]) {
                const int temp = arr[j];
                arr[j] = arr[j + 1];
                arr[j + 1] = temp;
                is_switched = true;
            }
        }
        if (!is_switched) {
            break;
        }
    }
}

File: algorithm/sort/BubbleSort2.java
package com.example;

import java.util.Arrays;

public class BubbleSort2 {
    /**
     * 원본 수정없이 새로운 배열 반환
     */
    public int[] bubbleSort(int[] arr) {
        int[] copiedArr = Arrays.copyOf(arr, arr.length);
        for (int i = arr.length - 1; i > 0; i--) {
            for (int j = 0; j < i; j++) {
                if (copiedArr[j] > copiedArr[j + 1]) {
                    switchElements(copiedArr, j, j + 1);
                }
            }
        }
        return copiedArr;
    }

    private void switchElements(int[] arr, int idx1, int idx2) {
        int temp = arr[idx1];
        arr[idx1] = arr[idx2];
        arr[idx2] = temp;
    }
}
